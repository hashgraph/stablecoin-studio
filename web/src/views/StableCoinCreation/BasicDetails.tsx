import { Heading, Stack, VStack } from '@chakra-ui/react';
import type { Control, FieldValues } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import InputController from '../../components/Form/InputController';

interface BasicDetailsProps {
	control: Control<FieldValues>;
}

const BasicDetails = (props: BasicDetailsProps) => {
	const { control } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	return (
		<VStack h='full' justify={'space-between'}>
			<VStack pt='80px'>
				<Stack minW={400}>
					<Heading
						data-testid='title'
						fontSize='24px'
						fontWeight='700'
						mb={10}
						lineHeight='16px'
						textAlign={'left'}
					>
						{t('stableCoinCreation:basicDetails.title')}
					</Heading>
					<Stack as='form' spacing={6}>
						<InputController
							rules={{
								required: t(`global:validations.required`),
							}}
							isRequired
							control={control}
							name={'name'}
							label={t('stableCoinCreation:basicDetails.name')}
							placeholder={t('stableCoinCreation:basicDetails.namePlaceholder')}
						/>
						<InputController
							rules={{
								required: t(`global:validations.required`),
							}}
							isRequired
							control={control}
							name={'symbol'}
							label={t('stableCoinCreation:basicDetails.symbol')}
							placeholder={t('stableCoinCreation:basicDetails.symbolPlaceholder')}
						/>
					</Stack>
				</Stack>
			</VStack>
		</VStack>
	);
};

export default BasicDetails;
