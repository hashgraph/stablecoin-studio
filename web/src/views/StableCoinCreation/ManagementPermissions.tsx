import { Heading, HStack, Stack, Text, VStack } from '@chakra-ui/react';
import type { Control, FieldValues } from 'react-hook-form';
import { useWatch } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import SwitchController from '../../components/Form/SwitchController';
import KeySelector from './components/KeySelector';

interface ManagementPermissionsProps {
	control: Control<FieldValues>;
}

const ManagementPermissions = (props: ManagementPermissionsProps) => {
	const { control } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	const isManagementPermissions = useWatch({
		control,
		name: 'managementPermissions',
	});

	const keys = [
		{
			name: 'adminKey',
			nameTranslate: t('stableCoinCreation:managementPermissions.admin'),
		},
		{
			name: 'supplyKey',
			nameTranslate: t('stableCoinCreation:managementPermissions.supply'),
		},
		{
			name: 'wipeKey',
			nameTranslate: t('stableCoinCreation:managementPermissions.wipe'),
		},
		{
			name: 'freezeKey',
			nameTranslate: t('stableCoinCreation:managementPermissions.freeze'),
		},
		{
			name: 'pauseKey',
			nameTranslate: t('stableCoinCreation:managementPermissions.pause'),
		},
	];

	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:managementPermissions.title')}
				</Heading>
				<Stack as='form' spacing={6} pb={6}>
					<HStack mb={4}>
						<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
							{t('stableCoinCreation:managementPermissions.description')}
						</Text>
						<SwitchController
							control={control}
							name={'managementPermissions'}
							defaultValue={true}
						/>
					</HStack>
					{isManagementPermissions === false ? (
						<>
							{keys.map((item) => {
								return (
									<KeySelector
										key={item.name}
										control={control}
										name={item.name}
										label={item.nameTranslate}
									/>
								);
							})}
						</>
					) : (
						<Stack spacing={2} fontSize='14px' fontWeight='400'>
							{keys.map((item, index) => {
								return <Text key={index}>{item.nameTranslate}</Text>;
							})}
						</Stack>
					)}
				</Stack>
			</Stack>
		</VStack>
	);
};

export default ManagementPermissions;
