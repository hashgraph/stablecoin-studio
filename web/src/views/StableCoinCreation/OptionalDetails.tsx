import { Heading, Stack, VStack } from '@chakra-ui/react';
import type { Control, FieldValues } from 'react-hook-form';
import { useWatch } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import InputNumberController from '../../components/Form/InputNumberController';
import { SelectController } from '../../components/Form/SelectController';

interface OptionalDetailsProps {
	control: Control<FieldValues>;
}

const OptionalDetails = (props: OptionalDetailsProps) => {
	const { control } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	const supplyTypes = [
		{
			value: 0,
			label: t('stableCoinCreation:optionalDetails.infinite'),
		},
		{
			value: 1,
			label: t('stableCoinCreation:optionalDetails.finite'),
		},
	];

	const selectorStyle = {
		wrapper: {
			border: '1px',
			borderColor: 'brand.black',
			borderRadius: '8px',
			height: 'min',
		},
		menuList: {
			maxH: '220px',
			overflowY: 'auto',
			bg: 'brand.white',
			boxShadow: 'down-black',
			p: 2,
			zIndex: 99,
		},
		valueSelected: {
			fontSize: '14px',
			fontWeight: '500',
		},
	};

	const isSupplyTypeFinite =
		useWatch({
			control,
			name: 'supplyType',
		})?.value === 1;

	const initialSupply = useWatch({
		control,
		name: 'initialSupply',
	});

	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:optionalDetails.title')}
				</Heading>
				<Stack as='form' spacing={6}>
					<InputNumberController
						rules={{
							required: t(`global:validations.required`),
						}}
						isRequired
						control={control}
						name={'initialSupply'}
						label={t('stableCoinCreation:optionalDetails.initialSupply')}
						placeholder={t('stableCoinCreation:optionalDetails.placeholder', {
							placeholder: t('stableCoinCreation:optionalDetails.initialSupply'),
						})}
					/>
					<SelectController
						control={control}
						name={'supplyType'}
						options={supplyTypes}
						label={t('stableCoinCreation:optionalDetails.typeSupply')}
						placeholder={t('stableCoinCreation:optionalDetails.typeSupplyPlaceholder')}
						overrideStyles={selectorStyle}
						addonLeft={true}
						variant='unstyled'
						defaultValue={'0'}
					/>
					{isSupplyTypeFinite && (
						<InputNumberController
							rules={{
								required: t(`global:validations.required`),
								validate: {
									quantityOverTotalSupply: (value: number) => {
										return (
											(initialSupply && initialSupply <= value) ||
											t('global:validations.overMaxSupply')
										);
									},
								},
							}}
							isRequired
							control={control}
							name={'maxSupply'}
							label={t('stableCoinCreation:optionalDetails.maxSupply')}
							placeholder={t('stableCoinCreation:optionalDetails.placeholder', {
								placeholder: t('stableCoinCreation:optionalDetails.maxSupply'),
							})}
						/>
					)}
					<InputNumberController
						rules={{
							required: t(`global:validations.required`),
						}}
						isRequired
						control={control}
						name={'decimals'}
						label={t('stableCoinCreation:optionalDetails.decimals')}
						placeholder={t('stableCoinCreation:optionalDetails.placeholder', {
							placeholder: t('stableCoinCreation:optionalDetails.decimals'),
						})}
						maxValue={18}
						initialValue={6}
					/>
				</Stack>
			</Stack>
		</VStack>
	);
};

export default OptionalDetails;
