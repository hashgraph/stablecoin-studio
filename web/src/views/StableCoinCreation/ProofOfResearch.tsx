import { Heading, HStack, Stack, Text, VStack } from '@chakra-ui/react';
import type { CreateRequest } from 'hedera-stable-coin-sdk';


import type { Control, FieldValues, UseFormReturn} from 'react-hook-form';
import { useWatch  } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import InputController from '../../components/Form/InputController';
import SwitchController from '../../components/Form/SwitchController';
// import { handleRequestValidation } from '../../utils/validationsHelper.js';


interface ProofOfResearchProps {
	form : UseFormReturn;
	control: Control<FieldValues>;
	request: CreateRequest;
}

const ProofOfResearch = (props: ProofOfResearchProps) => {
	const { control } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	const isProofOfResearch = useWatch({
		control,
		name: 'proofOfResearch',
		
	});
	
	/* useEffect(() => {
		if (!isProofOfResearch) {
			form.resetField('PoRInitialAmount', { defaultValue: "" });
			form.resetField('PoR', { defaultValue: "" });
		}
	}, [isProofOfResearch]);
	*/
	

	const haveDataFeed = useWatch({
	
		control,
		name: 'hasDataFeed',
	});
/*
	useEffect(() => {
		if (!haveDataFeed) {
			form.resetField('PoR', { defaultValue: "" });
		}else{
			form.resetField('PoRInitialAmount', { defaultValue: "" });
		}
	}, [haveDataFeed]);
*/	
	
	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:proofOfResearch.title')}
				</Heading>
			
				<HStack>
					<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
						{t('stableCoinCreation:proofOfResearch.description')}
					</Text>
					<SwitchController
						control={control}
						name={'proofOfResearch'}
						defaultValue={false}
					/>
				</HStack>

				{isProofOfResearch === true ? (<HStack >
						<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
							{t('stableCoinCreation:proofOfResearch.haveDataFeed')}
						</Text>
						<SwitchController
							control={control}
							name={'hasDataFeed'}
							defaultValue={false}
						/>
					</HStack>) : (<></>		
				)}


				{(haveDataFeed === true && isProofOfResearch === true)  ?  (
				<HStack >
						<InputController
						rules={{
							// required: t(`global:validations.required`),
							/* validate: {
								validation: (value: string) => {
									request.symbol = value;
									const res = handleRequestValidation(request.validate('symbol'));
									return res;
								},
							}, */
						}}
						
						control={control}
						name={'PoR'}
						label={t('stableCoinCreation:proofOfResearch.dataFeed')}
						placeholder={t('stableCoinCreation:proofOfResearch.dataFeedPlaceholder')}
						/>
					</HStack>
			

					
				):(<></>) }

				{ (haveDataFeed === false && isProofOfResearch === true)  ?  (
				
				<HStack >
					<InputController
						rules={{
							// required: t(`global:validations.required`),
							/* validate: {
								validation: (value: string) => {
									request.symbol = value;
									const res = handleRequestValidation(request.validate('symbol'));
									return res;
								},
							}, */
						}}
						
						control={control}
						name={'PoRInitialAmount'}
						label={t('stableCoinCreation:proofOfResearch.initialSupply')}
						placeholder={t('stableCoinCreation:proofOfResearch.initialSupply')}
						/>

				</HStack>
									
				):(<></>) }				
								
			</Stack>
		</VStack>
	);
};

export default ProofOfResearch;
