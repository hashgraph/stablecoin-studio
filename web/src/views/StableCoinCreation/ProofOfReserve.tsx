import { Heading, HStack, Stack, Text, VStack } from '@chakra-ui/react';
import type { CreateRequest } from 'hedera-stable-coin-sdk';
import { useEffect } from 'react';

import type { Control, FieldValues, UseFormReturn } from 'react-hook-form';
import { useWatch } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import InputController from '../../components/Form/InputController';
import SwitchController from '../../components/Form/SwitchController';
import { propertyNotFound } from '../../constant';
import { handleRequestValidation } from '../../utils/validationsHelper';

interface ProofOfReserveProps {
	form: UseFormReturn;
	control: Control<FieldValues>;
	request: CreateRequest;
}

const ProofOfReserve = (props: ProofOfReserveProps) => {
	const { control, form, request } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	const proofOfReserve = useWatch({
		control,
		name: 'proofOfReserve',
	});

	const hasDataFeed = useWatch({
		control,
		name: 'hasDataFeed',
	});

	useEffect(() => {
		if (!proofOfReserve) {
			form.resetField('reserveInitialAmount');
			form.resetField('reserveAddress');
		}
	}, [proofOfReserve]);

	useEffect(() => {
		if (!hasDataFeed) {
			form.resetField('reserveAddress');
		} else {
			form.resetField('reserveInitialAmount');
		}
	}, [hasDataFeed]);

	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:proofOfReserve.title')}
				</Heading>

				<HStack>
					<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
						{t('stableCoinCreation:proofOfReserve.description')}
					</Text>
					<SwitchController control={control} name={'proofOfReserve'} defaultValue={false} />
				</HStack>

				{proofOfReserve === true && (
					<HStack>
						<Text maxW={'252px'} fontSize='14px' fontWeight='400' lineHeight='17px'>
							{t('stableCoinCreation:proofOfReserve.haveDataFeed')}
						</Text>
						<SwitchController control={control} name={'hasDataFeed'} defaultValue={false} />
					</HStack>
				)}

				{hasDataFeed === true && proofOfReserve === true && (
					<HStack pt={'15px'}>
						<InputController
							rules={{
								required: t(`global:validations.required`) ?? propertyNotFound,
								validate: {
									validation: (value: string) => {
										request.createReserve = proofOfReserve;
										request.reserveAddress = value;
										const res = handleRequestValidation(request.validate('reserveAddress'));
										return res;
									},
								},
							}}
							isRequired
							control={control}
							name={'reserveAddress'}
							label={t('stableCoinCreation:proofOfReserve.dataFeed') ?? propertyNotFound}
							placeholder={
								t('stableCoinCreation:proofOfReserve.dataFeedPlaceholder') ?? propertyNotFound
							}
						/>
					</HStack>
				)}

				{!hasDataFeed && proofOfReserve === true && (
					<HStack pt={'15px'}>
						<InputController
							rules={{
								required: t(`global:validations.required`) ?? propertyNotFound,
								validate: {
									validation: (value: string) => {
										request.createReserve = proofOfReserve;
										request.reserveInitialAmount = value;
										const res = handleRequestValidation(request.validate('reserveInitialAmount'));
										return res;
									},
								},
							}}
							isRequired
							control={control}
							name={'reserveInitialAmount'}
							label={t('stableCoinCreation:proofOfReserve.initialSupply') ?? propertyNotFound}
							placeholder={t('stableCoinCreation:proofOfReserve.initialSupply') ?? propertyNotFound}
						/>
					</HStack>
				)}
			</Stack>
		</VStack>
	);
};

export default ProofOfReserve;
