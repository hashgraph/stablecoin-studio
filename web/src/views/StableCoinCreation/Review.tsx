import { Heading, Stack, VStack } from '@chakra-ui/react';
import type { UseFormReturn } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import DetailsReview from '../../components/DetailsReview';
import { OTHER_KEY_VALUE } from './components/KeySelector';

interface ReviewProps {
	form: UseFormReturn;
}

const Review = (props: ReviewProps) => {
	const { form } = props;
	const { t, i18n } = useTranslation(['global', 'stableCoinCreation']);

	const { getValues } = form;
	const {
		name,
		symbol,
		autorenewAccount,
		initialSupply,
		supplyType,
		totalSupply,
		decimals,
		expirationDate,
		managementPermissions,
		adminKey,
		supplyKey,
		rescueKey,
		wipeKey,
		freezeKey,
		feeScheduleKey,
		treasuryAccountAddress,
	} = getValues();

	const formatDate = (date = '') => {
		return new Date(date).toLocaleDateString(i18n.language, {
			year: 'numeric',
			month: '2-digit',
			day: '2-digit',
		});
	};

	const getKey = (keySelected: { value: number; label: string }, nameOtherKey: string) => {
		const { value, label } = keySelected;

		if (value === OTHER_KEY_VALUE) {
			return label + ': ' + form.watch(nameOtherKey);
		}

		return label;
	};

	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:review.title')}
				</Heading>
				<Stack as='form' spacing={9} pb='50px'>
					<DetailsReview
						title={t('stableCoinCreation:review.basicDetails')}
						titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
						details={[
							{
								label: t('stableCoinCreation:basicDetails.name'),
								value: name || '',
							},
							{
								label: t('stableCoinCreation:basicDetails.symbol'),
								value: symbol || '',
							},
							{
								label: t('stableCoinCreation:basicDetails.autorenewAccount'),
								value: autorenewAccount || '',
							},
						]}
					/>
					<DetailsReview
						title={t('stableCoinCreation:review.optionalDetails')}
						titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
						details={[
							{
								label: t('stableCoinCreation:optionalDetails.initialSupply'),
								value: initialSupply || '',
							},
							{
								label: t('stableCoinCreation:optionalDetails.typeSupply'),
								value: supplyType?.label || '',
							},
							{
								label: t('stableCoinCreation:optionalDetails.totalSupply'),
								value: supplyType?.value === 1 ? totalSupply : supplyType?.label,
							},
							{
								label: t('stableCoinCreation:optionalDetails.decimals'),
								value: decimals || '',
							},
							{
								label: t('stableCoinCreation:optionalDetails.expirationDate'),
								value: expirationDate ? formatDate(expirationDate) : '',
							},
						]}
					/>
					<DetailsReview
						title={t('stableCoinCreation:review.managementPermissions')}
						titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
						details={[
							{
								label: t('stableCoinCreation:managementPermissions.stableCoinAdmin'),
								value: managementPermissions
									? t('stableCoinCreation:managementPermissions.theSmartContract')
									: getKey(adminKey, 'adminKeyOther'),
							},
							{
								label: t('stableCoinCreation:managementPermissions.supply'),
								value: managementPermissions
									? t('stableCoinCreation:managementPermissions.theSmartContract')
									: getKey(supplyKey, 'supplyKeyOther'),
							},
							{
								label: t('stableCoinCreation:managementPermissions.rescue'),
								value: managementPermissions
									? t('stableCoinCreation:managementPermissions.theSmartContract')
									: getKey(rescueKey, 'rescueKeyOther'),
							},
							{
								label: t('stableCoinCreation:managementPermissions.wipe'),
								value: managementPermissions
									? t('stableCoinCreation:managementPermissions.theSmartContract')
									: getKey(wipeKey, 'wipeKeyOther'),
							},
							{
								label: t('stableCoinCreation:managementPermissions.freeze'),
								value: managementPermissions
									? t('stableCoinCreation:managementPermissions.theSmartContract')
									: getKey(freezeKey, 'freezeKeyOther'),
							},
							{
								label: t('stableCoinCreation:managementPermissions.feeSchedule'),
								value: managementPermissions
									? t('stableCoinCreation:managementPermissions.theSmartContract')
									: getKey(feeScheduleKey, 'feeScheduleKeyOther'),
							},
						]}
					/>
					<DetailsReview
						title={t('stableCoinCreation:managementPermissions.treasuryAccountAddress')}
						titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
						details={[
							{
								label: t('stableCoinCreation:managementPermissions.treasuryAccountAddress'),
								value:
									!managementPermissions && supplyKey.value === OTHER_KEY_VALUE
										? treasuryAccountAddress
										: t('stableCoinCreation:managementPermissions.theSmartContract'),
							},
						]}
					/>
				</Stack>
			</Stack>
		</VStack>
	);
};

export default Review;
