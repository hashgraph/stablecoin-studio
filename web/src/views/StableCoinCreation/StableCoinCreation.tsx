import { Box, Stack } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';
import { useForm } from 'react-hook-form';
import BaseContainer from '../../components/BaseContainer';
import BasicDetails from './BasicDetails';
import type { Step } from '../../components/Stepper';
import Stepper from '../../components/Stepper';
import { NamedRoutes } from '../../Router/NamedRoutes';
import { RouterManager } from '../../Router/RouterManager';

const StableCoinCreation = () => {
	const navigate = useNavigate();
	const { t } = useTranslation('stableCoinCreation');
	const {
		control,
		formState: { isValid },
	} = useForm({ mode: 'onChange' });

	// TODO: change childrens props when components will be ready
	const steps: Step[] = [
		{
			number: '01',
			title: t('tabs.basicDetails'),
			complete: false,
			children: <BasicDetails control={control} />,
		},
		{
			number: '02',
			title: t('tabs.optionalDetails'),
			complete: false,
			children: <Box>Tab 2</Box>,
		},
		{
			number: '03',
			title: t('tabs.managementPermissions'),
			complete: false,
			children: <Box>Tab 3</Box>,
		},
		{
			number: '04',
			title: t('tabs.review'),
			complete: false,
			children: <Box>Tab 4</Box>,
		},
	];

	const handleFinish = () => {
		// TODO: connect with SDK services
		alert('create!');
	};

	const handleCancel = () => {
		RouterManager.to(navigate, NamedRoutes.Dashboard);
	};

	const stepperProps = {
		steps,
		handleLastButtonPrimary: handleFinish,
		handleFirstButtonSecondary: handleCancel,
		textLastButtonPrimary: t('common.createStableCoin'),
		isValid,
	};

	return (
		<Stack h='full'>
			<BaseContainer title={t('common.createNewStableCoin')}>
				<Stepper {...stepperProps} />
			</BaseContainer>
		</Stack>
	);
};

export default StableCoinCreation;
