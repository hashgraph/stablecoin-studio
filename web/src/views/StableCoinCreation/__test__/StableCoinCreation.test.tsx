import { render } from '../../../test/index';
import translations from '../../../translations/en/stableCoinCreation.json';
import StableCoinCreation from '../StableCoinCreation';
import configureMockStore from 'redux-mock-store';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@testing-library/react';
import SDKService from '../../../services/SDKService';
import ContractId from '@hashgraph/stablecoin-npm-sdk/build/esm/src/domain/context/contract/ContractId';
import { SupportedWallets } from '@hashgraph/stablecoin-npm-sdk';
import { HederaId } from '@hashgraph/stablecoin-npm-sdk/build/esm/src/domain/context/shared/HederaId';

describe(`<${StableCoinCreation.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<StableCoinCreation />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<StableCoinCreation />);

		const header = component.getByTestId('creation-title');
		expect(header).toHaveTextContent(translations.common.createNewStableCoin);
	});

	test('should has subtitle', () => {
		const component = render(<StableCoinCreation />);

		const subtitle = component.getByTestId('creation-subtitle');
		expect(subtitle).toHaveTextContent(translations.common.factoryId);
	});

	test('should have options', async () => {
		const mockStore = configureMockStore();
		const store = mockStore({
			wallet: {
				lastWallet: SupportedWallets.BLADE,
				factoryId: '0.0.12345',
				accountInfo: {
					id: '0.0.12345',
				},
				data: {
					account: {
						id: '0.0.12345',
					},
				},
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					hederaTokenManagerId: '0.0.12345',
				}),
				watch: () => jest.fn(),
			}),
		}));

		const contractId: ContractId = new ContractId('0.0.1234');
		jest.spyOn(SDKService, 'getHederaTokenManagerList').mockResolvedValue([contractId]);

		const createResponse = {
			coin: { tokenId: new HederaId('0.0.12345') },
			reserve: { proxyAddress: new ContractId('0.0.1234') },
		};
		const createResponse2 = {};
		jest.spyOn(SDKService, 'createStableCoin').mockResolvedValue(createResponse);

		const component = render(<StableCoinCreation />, store);

		const noProof = component.getByTestId('no-proof-of-reserve-title');
		await waitFor(() => {
			expect(noProof).not.toBeInTheDocument();
		});

		//step 1
		const name = component.getByTestId('name');
		await userEvent.type(name, 'name');

		const symbol = component.getByTestId('symbol');
		await userEvent.type(symbol, 'symbol');

		const next1 = component.getByTestId('stepper-step-panel-button-primary-1');
		await userEvent.click(next1);

		//step 2
		await waitFor(() => {
			const initialSupply = component.getByTestId('initialSupply');
			userEvent.type(initialSupply, '1000');
		});

		const select = component.getByTestId('select-placeholder');
		await userEvent.click(select);

		const next = component.getByTestId('stepper-step-panel-button-primary-2');
		await userEvent.click(next);

		//step 3
		await waitFor(() => {
			const next = component.getByTestId('stepper-step-panel-button-primary-3');
			userEvent.click(next);
		});

		//step 4
		await waitFor(() => {
			const next = component.getByTestId('stepper-step-panel-button-primary-4');
			userEvent.click(next);
		});

		//step 5
		await waitFor(() => {
			const title = component.getByText('Create Stablecoin');
			userEvent.click(title);
		});

		const nextLast = component.getByTestId('stepper-step-panel-button-primary-5');
		await userEvent.click(nextLast);
	});
});
