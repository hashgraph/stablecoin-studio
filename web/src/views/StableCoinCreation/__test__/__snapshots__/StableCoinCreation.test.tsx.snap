// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StableCoinCreation /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-19t9j1"
  >
    <div
      class="chakra-stack css-1u2nxl"
    >
      <p
        class="chakra-text css-ai2tcq"
        data-testid="creation-title"
      >
        Create new stablecoin
      </p>
      <i
        class="chakra-text css-zcexcl"
        data-testid="creation-subtitle"
      >
        Factory Id : undefined
      </i>
    </div>
    <div
      class="chakra-stack css-19t9j1"
    >
      <h2
        class="chakra-heading css-bqsusc"
        data-testid="base-container-heading"
      />
      <div
        class="css-1b6lda7"
        data-testid="base-container-children"
      >
        <div
          class="chakra-tabs css-1x6vjw3"
        >
          <div
            aria-orientation="horizontal"
            class="chakra-tabs__tablist css-9jay18"
            role="tablist"
          >
            <button
              aria-controls="tabs-:r1:--tabpanel-0"
              aria-disabled="false"
              aria-selected="true"
              class="chakra-tabs__tab css-h9y905"
              data-index="0"
              id="tabs-:r1:--tab-0"
              role="tab"
              tabindex="0"
              type="button"
            >
              <p
                class="chakra-text css-1ikrtdg"
                data-testid="stepper-step-number-1"
              >
                01
              </p>
              <p
                class="chakra-text css-1i91963"
                data-testid="stepper-step-title-1"
              >
                Stablecoin Details
              </p>
            </button>
            <button
              aria-controls="tabs-:r1:--tabpanel-1"
              aria-selected="false"
              class="chakra-tabs__tab css-h9y905"
              data-index="1"
              disabled=""
              id="tabs-:r1:--tab-1"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1wzl1bo"
                data-testid="stepper-step-number-2"
              >
                02
              </p>
              <p
                class="chakra-text css-1i91963"
                data-testid="stepper-step-title-2"
              >
                Supply & Metadata
              </p>
            </button>
            <button
              aria-controls="tabs-:r1:--tabpanel-2"
              aria-selected="false"
              class="chakra-tabs__tab css-h9y905"
              data-index="2"
              disabled=""
              id="tabs-:r1:--tab-2"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1wzl1bo"
                data-testid="stepper-step-number-3"
              >
                03
              </p>
              <p
                class="chakra-text css-1i91963"
                data-testid="stepper-step-title-3"
              >
                Configure Permissions
              </p>
            </button>
            <button
              aria-controls="tabs-:r1:--tabpanel-3"
              aria-selected="false"
              class="chakra-tabs__tab css-h9y905"
              data-index="3"
              disabled=""
              id="tabs-:r1:--tab-3"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1wzl1bo"
                data-testid="stepper-step-number-4"
              >
                04
              </p>
              <p
                class="chakra-text css-1i91963"
                data-testid="stepper-step-title-4"
              >
                Proof-Of-Reserve
              </p>
            </button>
            <button
              aria-controls="tabs-:r1:--tabpanel-4"
              aria-selected="false"
              class="chakra-tabs__tab css-h9y905"
              data-index="4"
              disabled=""
              id="tabs-:r1:--tab-4"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1wzl1bo"
                data-testid="stepper-step-number-5"
              >
                05
              </p>
              <p
                class="chakra-text css-1i91963"
                data-testid="stepper-step-title-5"
              >
                Create Stablecoin
              </p>
            </button>
          </div>
          <div
            class="chakra-tabs__tab-panels css-27a8z"
          >
            <div
              aria-labelledby="tabs-:r1:--tab-0"
              class="chakra-tabs__tab-panel css-1iyuyap"
              data-testid="stepper-step-panel-container-1"
              id="tabs-:r1:--tabpanel-0"
              role="tabpanel"
              tabindex="0"
            >
              <div
                class="chakra-stack css-tc011i"
              >
                <div
                  class="css-1c3xk1u"
                >
                  <div
                    class="chakra-stack css-1wbej9n"
                  >
                    <div
                      class="css-10opskl"
                      data-testid="no-proof-of-reserve-container"
                    >
                      <div
                        class="css-mnqf94"
                      >
                        <svg
                          viewBox="0 0 2500 2500"
                          width="48px"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <g
                            data-name="Layer 2"
                          >
                            <g
                              data-name="Layer 1"
                            >
                              <path
                                d="M1250 0C559.64 0 0 559.64 0 1250s559.64 1250 1250 1250 1250-559.64 1250-1250S1940.36 0 1250 0"
                              />
                              <path
                                d="M1758.12 1790.62h-158.74v-337.49H900.62v337.49H741.87V696.25h158.75v329.37h698.76V696.25h158.75zm-850-463.75h698.75V1152.5H908.12z"
                                fill="#fff"
                              />
                            </g>
                          </g>
                        </svg>
                      </div>
                      <p
                        class="chakra-text css-e8j0ix"
                        data-testid="no-proof-of-reserve-title"
                      >
                        Loading data
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="css-1q6t3f6"
                >
                  <hr
                    aria-orientation="horizontal"
                    class="chakra-divider css-12thghs"
                  />
                  <div
                    class="css-1e334ws"
                  >
                    <button
                      class="chakra-button css-hi7zxp"
                      data-testid="stepper-step-panel-button-secondary-1"
                      type="button"
                    >
                      Cancel
                    </button>
                    <button
                      class="chakra-button css-qvkb6r"
                      data-testid="stepper-step-panel-button-primary-1"
                      disabled=""
                      type="button"
                    >
                      Next step
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="tabs-:r1:--tab-1"
              class="chakra-tabs__tab-panel css-1iyuyap"
              data-testid="stepper-step-panel-container-2"
              hidden=""
              id="tabs-:r1:--tabpanel-1"
              role="tabpanel"
              tabindex="0"
            />
            <div
              aria-labelledby="tabs-:r1:--tab-2"
              class="chakra-tabs__tab-panel css-1iyuyap"
              data-testid="stepper-step-panel-container-3"
              hidden=""
              id="tabs-:r1:--tabpanel-2"
              role="tabpanel"
              tabindex="0"
            />
            <div
              aria-labelledby="tabs-:r1:--tab-3"
              class="chakra-tabs__tab-panel css-1iyuyap"
              data-testid="stepper-step-panel-container-4"
              hidden=""
              id="tabs-:r1:--tabpanel-3"
              role="tabpanel"
              tabindex="0"
            />
            <div
              aria-labelledby="tabs-:r1:--tab-4"
              class="chakra-tabs__tab-panel css-1iyuyap"
              data-testid="stepper-step-panel-container-5"
              hidden=""
              id="tabs-:r1:--tabpanel-4"
              role="tabpanel"
              tabindex="0"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
